File: appboxd/lib/watermark.dart
/// Post-emit provenance/watermark pass — pure-Dart port of the archived
/// tools/watermark/watermark.mjs.
///
/// Walks an emitted tree and prepends a provenance block (`appbox:provenance`
/// marker) to every `.dart/.js/.mjs/.ts/.yaml/.yml/.sh/.html` file. Free tier
/// carries a watermark line inside the block; paid tier emits a clean
/// provenance header. `.json` has no comment syntax — its provenance is a
/// sibling entry in the manifest instead. Writes `.appbox-provenance.json`
/// (a sha256 manifest of every touched file, post-injection).
///
/// The block is EVIDENCE, not a lock — stripping it is handled by licence
/// terms, not crypto (Unity Personal splash precedent). Emitted source is
/// never encrypted. Idempotent via the [marker] token.
///
/// Licence status comes from `licence.dart` (LicenceVerdict.unlocks). Only
/// `paid` (valid/grace) is clean; `none`/invalid/expired all watermark like
/// `free`, never blocking emission.
library;

import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:path/path.dart' as p;

import 'crypto_aead.dart';
import 'licence.dart';

/// The idempotency token every provenance block carries.
const marker = 'appbox:provenance';

/// Watermark line emitted inside the block on the free tier.
const watermarkLine = 'Built with appbox (free tier) — https://appbox.dev';

/// Name of the sha256 manifest written at the tree root.
const manifestName = '.appbox-provenance.json';

/// Comment syntax per extension. `.html` is special (HTML comment); `.json`
/// is deliberately absent — no comment syntax.
const commentSyntax = <String, String>{
  '.dart': '//',
  '.js': '//',
  '.mjs': '//',
  '.ts': '//',
  '.yaml': '#',
  '.yml': '#',
  '.sh': '#',
  '.html': 'html',
};

final _doctype = RegExp(r'^\s*<!doctype[^>]*>', caseSensitive: false);

/// Renders the provenance block for [ext] (lowercased, with leading dot).
String renderBlock(String ext, {required String tier, required String projectHash}) {
  final lines = <String>[
    marker,
    'generator: appbox  licence: $tier  project: $projectHash',
  ];
  if (tier == 'free') lines.add(watermarkLine);
  if (ext == '.html') {
    // <!-- appbox:provenance
    //      generator: ...
    //      Built with appbox (free tier) — ... -->
    final body = lines.skip(1).map((l) => '     $l').join('\n');
    return '<!-- ${lines.first}\n$body\n-->\n';
  }
  final c = commentSyntax[ext]!;
  return '${lines.map((l) => '$c $l').join('\n')}\n';
}

/// Prepends a provenance header to [filePath]. Returns one of:
/// `injected` | `already` | `json-skip` | `unsupported`.
String injectProvenance(String filePath,
    {required String tier, required String projectHash}) {
  final ext = p.extension(filePath).toLowerCase();
  if (ext == '.json') return 'json-skip';
  if (!commentSyntax.containsKey(ext)) return 'unsupported';

  final src = File(filePath).readAsStringSync();
  if (src.contains(marker)) return 'already';

  final block = renderBlock(ext, tier: tier, projectHash: projectHash);
  String out;
  if (src.startsWith('#!')) {
    // Shebang must stay line 1.
    final nl = src.indexOf('\n');
    out = src.substring(0, nl + 1) + block + src.substring(nl + 1);
  } else if (ext == '.html' && _doctype.hasMatch(src)) {
    // A comment before <!DOCTYPE> triggers quirks mode — inject after it.
    out = src.replaceFirstMapped(_doctype, (m) => '${m[0]}$block');
  } else {
    out = block + src;
  }
  File(filePath).writeAsStringSync(out);
  return 'injected';
}

/// Walks [rootDir], injects provenance into every supported file, and writes
/// `.appbox-provenance.json` (`{tier, ts, files:[{path, sha256, note?}]}`).
///
/// [tier] overrides licence detection (`'paid'`/`'free'`); when null it
/// resolves from `licence.dart` (valid/grace → paid, else free). [licencePath]
/// overrides the licence file location. Returns the manifest map.
Map<String, Object?> runPass(String rootDir,
    {String? tier, String? licencePath, String? projectHash}) {
  final resolvedTier = tier ?? _resolveTier(licencePath);
  final hash = projectHash ??
      hexEncode(sha256(Uint8List.fromList(utf8.encode(p.canonicalize(rootDir)))))
          .substring(0, 12);

  final files = <Map<String, Object?>>[];
  final entries = Directory(rootDir).listSync(recursive: true).toList()
    ..sort((a, b) => a.path.compareTo(b.path));
  for (final entity in entries) {
    if (entity is! File) continue;
    final segments = p.split(entity.path);
    if (segments.any((s) => s == 'node_modules' || s == '.git')) continue;
    if (p.basename(entity.path) == manifestName) continue;

    final ext = p.extension(entity.path).toLowerCase();
    if (!commentSyntax.containsKey(ext) && ext != '.json') continue;

    final result =
        injectProvenance(entity.path, tier: resolvedTier, projectHash: hash);
    final entry = <String, Object?>{
      'path': p.relative(entity.path, from: rootDir),
      'sha256': hexEncode(sha256(entity.readAsBytesSync())),
    };
    if (result == 'json-skip') {
      entry['note'] =
          'json: no comment syntax — provenance is this manifest entry';
    }
    files.add(entry);
  }

  final manifest = <String, Object?>{
    'tier': resolvedTier,
    'ts': DateTime.now().toUtc().toIso8601String(),
    'files': files,
  };
  File(p.join(rootDir, manifestName))
      .writeAsStringSync('${const JsonEncoder.withIndent('  ').convert(manifest)}\n');
  return manifest;
}

/// Mirrors licence_tool's path resolution: `~/.appbox/licence.json` then a
/// walk-up for `<repo>/pipeline/state/licence.json`. Empty string = absent.
String _defaultLicencePath() {
  final home = Platform.environment['HOME'] ?? Platform.environment['USERPROFILE'];
  if (home != null) {
    final userPath = p.join(home, '.appbox', 'licence.json');
    if (File(userPath).existsSync()) return userPath;
  }
  var root = Directory.current.path;
  while (!File(p.join(root, 'pipeline', 'pipeline.sh')).existsSync()) {
    final parent = Directory(root).parent.path;
    if (parent == root) return '';
    root = parent;
  }
  final repoPath = p.join(root, 'pipeline', 'state', 'licence.json');
  return File(repoPath).existsSync() ? repoPath : '';
}

/// Only a valid/grace licence is `paid`; everything else watermarks as `free`.
String _resolveTier(String? licencePath) {
  final path = licencePath ?? _defaultLicencePath();
  if (path.isEmpty) return 'free';
  return Licence.verifyFile(path).unlocks ? 'paid' : 'free';
}

File: tools/watermark/watermark.mjs
#!/usr/bin/env node
// tools/watermark/watermark.mjs — post-emit provenance/watermark pass.
//
// P1 (docs/plans/appbox-memory-and-payment.md): free tier = full product with
// watermarked outputs; paid = clean outputs. The provenance block is
// EVIDENCE, not a lock — stripping it is an accepted design constant handled
// by licence terms, not crypto (Unity Personal splash precedent,
// docs/research/monetization-and-licensing.md). Emitted source is never
// encrypted.
//
// Licence status: `dart run bin/licence_tool.dart status` inside appboxd,
// contract = JSON stdout {"status":"paid"|"free"|"none", tier, expires}.
// Any failure (tool missing, dart missing, bad JSON) => treated as "free".
// Status is cached per process.
//
// Env overrides (documented test hooks):
//   APPBOX_LICENCE_STATUS — JSON string or a plain status word ("paid"/"free");
//                           bypasses the dart call entirely.
//   APPBOXD_DIR           — where appboxd lives (default: <repo>/appboxd).
//
// CLI: node tools/watermark/watermark.mjs <rootDir>
import { createHash } from 'node:crypto';
import { execFileSync } from 'node:child_process';
import { readdirSync, readFileSync, statSync, writeFileSync } from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..', '..');

export const MARKER = 'appbox:provenance';
export const WATERMARK_LINE = 'Built with appbox (free tier) — https://appbox.dev';
export const MANIFEST_NAME = '.appbox-provenance.json';

// Comment syntax per extension. .json is deliberately absent: no comment
// syntax — its provenance is a sibling note in the manifest instead.
const COMMENT = {
  '.dart': '//', '.js': '//', '.mjs': '//', '.ts': '//',
  '.yaml': '#', '.yml': '#', '.sh': '#',
  '.html': 'html',
};

let _licence = null;
export function licenceStatus() {
  if (_licence) return _licence;
  const env = process.env.APPBOX_LICENCE_STATUS;
  if (env) {
    _licence = env.trim().startsWith('{')
      ? JSON.parse(env)
      : { status: env.trim() };
    return _licence;
  }
  try {
    const appboxd = process.env.APPBOXD_DIR || path.join(ROOT, 'appboxd');
    const out = execFileSync('dart', ['run', 'bin/licence_tool.dart', 'status'], {
      cwd: appboxd, encoding: 'utf8', stdio: ['ignore', 'pipe', 'ignore'],
    });
    _licence = JSON.parse(out);
  } catch {
    _licence = { status: 'free' }; // tool missing/broken => free, never blocks emission
  }
  return _licence;
}

// Contract: {status: "paid"|"free"|"none", ...}. Only "paid" is clean;
// "none" (no licence file) watermarks like "free".
function resolveTier(lic) {
  return (lic && lic.status) === 'paid' ? 'paid' : 'free';
}

function renderBlock(ext, { tier, projectHash }) {
  const lines = [
    MARKER,
    `generator: appbox  licence: ${tier}  project: ${projectHash}`,
    // evidence of origin, not a lock — see file header comment
  ];
  if (tier === 'free') lines.push(WATERMARK_LINE);
  if (ext === '.html') {
    return `<!-- ${lines[0]}\n${lines.slice(1).map((l) => `     ${l}`).join('\n')}\n-->\n`;
  }
  const c = COMMENT[ext];
  return lines.map((l) => `${c} ${l}`).join('\n') + '\n';
}

// Prepends a provenance header; free tier carries the watermark line inside
// the block. Idempotent via the MARKER token. Returns one of:
// 'injected' | 'already' | 'json-skip' | 'unsupported'.
export function injectProvenance(filePath, { tier, projectHash }) {
  const ext = path.extname(filePath).toLowerCase();
  if (ext === '.json') return 'json-skip';
  if (!(ext in COMMENT)) return 'unsupported';
  const src = readFileSync(filePath, 'utf8');
  if (src.includes(MARKER)) return 'already';
  const block = renderBlock(ext, { tier, projectHash });
  let out;
  if (src.startsWith('#!')) {
    // shebang must stay line 1
    const nl = src.indexOf('\n');
    out = src.slice(0, nl + 1) + block + src.slice(nl + 1);
  } else if (ext === '.html' && /^\s*<!doctype[^>]*>/i.test(src)) {
    // a comment before <!DOCTYPE> triggers quirks mode — inject after it
    out = src.replace(/^\s*<!doctype[^>]*>\s*/i, (m) => m + block);
  } else {
    out = block + src;
  }
  writeFileSync(filePath, out);
  return 'injected';
}

const sha256 = (buf) => createHash('sha256').update(buf).digest('hex');

// Walk rootDir, inject provenance into every supported file, and write
// .appbox-provenance.json {files: [{path, sha256, note?}], tier, ts}.
export function runPass(rootDir, opts = {}) {
  const tier = opts.tier || resolveTier(licenceStatus());
  const projectHash = opts.projectHash ||
    sha256(path.resolve(rootDir)).slice(0, 12);
  const files = [];

  (function walk(dir) {
    for (const name of readdirSync(dir).sort()) {
      if (name === 'node_modules' || name === '.git' || name === MANIFEST_NAME) continue;
      const p = path.join(dir, name);
      if (statSync(p).isDirectory()) { walk(p); continue; }
      const ext = path.extname(name).toLowerCase();
      if (!(ext in COMMENT) && ext !== '.json') continue;
      const result = injectProvenance(p, { tier, projectHash });
      const entry = { path: path.relative(rootDir, p), sha256: sha256(readFileSync(p)) };
      if (result === 'json-skip') entry.note = 'json: no comment syntax — provenance is this manifest entry';
      files.push(entry);
    }
  })(rootDir);

  const manifest = { tier, ts: new Date().toISOString(), files };
  writeFileSync(path.join(rootDir, MANIFEST_NAME), JSON.stringify(manifest, null, 2) + '\n');
  return manifest;
}

if (process.argv[1] && fileURLToPath(import.meta.url) === path.resolve(process.argv[1])) {
  const root = process.argv[2];
  if (!root) {
    console.error('usage: watermark.mjs <rootDir>');
    process.exit(64);
  }
  const m = runPass(root);
  console.log(`provenance pass: tier=${m.tier} files=${m.files.length} -> ${path.join(root, MANIFEST_NAME)}`);
}
